src/bin/ipa_bench/models.rs

```diff
@@ -109,57 +111,169 @@ impl SecretSharable for u64 {
     }
 }
 
-#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
-pub struct Event {
-    /// Secret shared and then encrypted match keys.
-    pub matchkeys: Vec<SecretShare>,
+pub trait EpochDuration: Sized + Add<Output = Self> + AddAssign + PartialOrd + From<u64> {
+    type Epoch;
+    type Offset;
 
-    /// The epoch which this event is generated. Using an 8-bit value = 256 epochs > 4 years (assuming 1 epoch = 1 week).
-    /// This field is in the clear.
-    pub epoch: u8,
+    const DAYS_IN_EPOCH: Self::Epoch;
+    const SECONDS_IN_EPOCH: u64;
+    const MAX: u64;
 
-    /// An offset in seconds into a given epoch. The clear is u32 (< 2^20 seconds), then encrypted and secret shared.
-    pub timestamp: SecretShare,
+    fn epoch(&self) -> Self::Epoch;
+    fn offset(&self) -> Self::Offset;
 }
 
-#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
-pub struct SourceEvent {
-    pub event: Event,
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct EventTimestamp(u64);
```

Comment:
+1 on hiding the internal representation of the timestamp and defining `From`!
also curious why we need `u64` if we seem to be able to keep everything inside a `u32`?

Reply:
I chose `u64` because rust's `Duration` struct uses it to represent seconds like below. It eliminates extra casts and can directly call `from`/`into`.

That said, the parameter is nameless and private, hence the only way to instantiate `EventTimestamp` outside `models` is `From`, which is defined @161 in this file.

```
pub struct Duration {
    secs: u64,
    nanos: u32,
}
```

Reply:
to me it makes sense to use the smallest container value possible - that helps a lot with CPU cache efficiency, memory usage etc. usually implementing `From<Duration>`, `From<u32>`, `From<u64>` is enough to make API easy to use